src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -1059,5 +1059,174 @@ public unsafe void CopyFromAscii(string data)
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        true,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    return new string(chars, 0, charIndex + charsUsed);
+                }
+                else if (block.Next == null)
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        following,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        true,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    return new string(chars, 0, charIndex + charsUsed);
+                }
+                else
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        following,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        false,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    charIndex += charsUsed;
+                    remaining -= following;
+                    block = block.Next;
+                    index = block.Start;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ArraySegment<byte> GetArraySegment(MemoryPoolIterator end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return default(ArraySegment<byte>);
+            }
+            if (end.Block == Block)
+            {
+                return new ArraySegment<byte>(Block.Array, Index, end.Index - Index);
+            }
+
+            return GetArraySegmentMultiBlock(ref end);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ArraySegment<byte> GetArraySegmentMultiBlock(ref MemoryPoolIterator end)
+        {
+            var length = GetLength(end);
+            var array = new byte[length];
+            CopyTo(array, 0, length, out length);
+            return new ArraySegment<byte>(array, 0, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector<byte> GetVector(byte vectorByte)
+        {
+            // Vector<byte> .ctor doesn't become an intrinsic due to detection issue
+            // However this does cause it to become an intrinsic (with additional multiply and reg->reg copy)
+            // https://github.com/dotnet/coreclr/issues/7459#issuecomment-253965670
+            return Vector.AsVectorByte(new Vector<ulong>(vectorByte * 0x0101010101010101ul));
```

Comment:
It was because `new Vector<byte>` wasn't recognised as an intrinsic so took a non-inlinable, very slow path https://github.com/dotnet/coreclr/issues/7459

This should have been resolved by https://github.com/dotnet/coreclr/pull/7683 in 1.2.0+ (and hopefully 4.6.3+ of the full framework?) - though this should work on all released versions of RyuJIT.

Both approaches still have a load to register step; the advantage of constructing the Vector in func are two fold
* Vector is created aligned so can use aligned cpu instructions; whereas byref is "indeterminate" so can't have as many optimizations applied
* It doesn't need to extra the byte from the vector for the byte search, which was suboptimal for Vectors (again fixed in 1.2.0+ https://github.com/dotnet/coreclr/pull/7329) and definately suboptimal for non-accelerated